mountinfo/mounted_linux.go

```diff
@@ -0,0 +1,82 @@
+package mountinfo
+
+import (
+	"os"
+	"path/filepath"
+
+	"golang.org/x/sys/unix"
+)
+
+func mountedByOpenat2(path string, st *unix.Stat_t) (bool, error) {
+	// This method works for all mounts (incl. bind mounts),
+	// but requires a recent (v5.6+) linux kernel.
+	var dir, file string
+
+	ftype := st.Mode & unix.S_IFMT
+	switch ftype {
+	case unix.S_IFDIR:
+		dir = path
+		file = ".."
+	default:
+		dir, file = filepath.Split(path)
+	}
```

Comment:
If you need to open the directory separately anyway, you can implement both directory and regular file handling as `openat2(dfd, "file", ...)` -- though I guess it's the same either way.

Reply:
I think that is what I'm doing (or I don't get what you mean here).

Reply:
I meant you could implement both cases (file and directory target) with `dir, file := filepath.Split(path)`. The `dir, file := path, ".."` trick I mentioned originally is an optimisation you can do if you already have the directory open, but if you don't have it open then both methods are equivalent.

But this is more of a nit than anything else, both ways work.

Reply:
That's right, not sure how I missed it earlier. Will update.

Reply:
Updated, simplified the code a lot!